File: P1/Ejercicio3/src/constructor/ConstructorBicicletaDecorada.java
package constructor;

import bicicleta.Bicicleta;
import decorador.DecoradorBicicletaConEstampado;
import decorador.DecoradorBicicletaConFunda;

/**
 * Clase constructora para bicicletas decoradas. Esta hereda de Constructor
 */
public class ConstructorBicicletaDecorada extends Constructor {
    /**
     * Constructor de la clase con parámetro. Inicializa la bicicleta a una bicicleta pasada por parámetro. Esto
     * permite construir cualquier tipo de bicicleta con decoración
     */
    public ConstructorBicicletaDecorada(Bicicleta bicicleta) {
        this.bicicleta = bicicleta;
    }

    /**
     * Agrega un estampado a la bicicleta actual
     */
    public void agregarEstampado() {
        bicicleta = new DecoradorBicicletaConEstampado(bicicleta);
    }

    /**
     * Agrega una funda a la bicicleta actual
     */
    public void agregarFunda() {
        bicicleta = new DecoradorBicicletaConFunda(bicicleta);
    }

    /**
     * Implementación del método abstracto 'recomponer', que reinicializa la instancia de bicicleta a 'null' (puesto
     * que puede ser cualquier tipo de bicicleta)
     */
    @Override
    public void recomponer() {
        bicicleta = null;
    }
}

File: P2/lib/Constructores/constructorBicicletaDecorada.dart
import 'constructor.dart';
import '../Bicicleta/Bicicleta.dart';
import '../Decorador/decoradorBicicletaConEstampado.dart';
import '../Decorador/decoradorBicicletaConFunda.dart';


class ConstructorBicicletaDecorada extends Constructor {
  ConstructorBicicletaDecorada(Bicicleta bicicleta) {
    this.bicicleta = bicicleta;
  }

  void agregarEstampado() {
    if (bicicleta != null) {
      bicicleta = DecoradorBicicletaConEstampado(bicicleta!);
    }
  }

  void agregarFunda() {
    if (bicicleta != null) {
      bicicleta = DecoradorBicicletaConFunda(bicicleta!);
    }
  }

  @override
  void recomponer() {
    bicicleta = null;
  }
}
